### PR TITLE
bug on tracks 'color_by' attribute 

### PR DESCRIPTION
**Pull request by Kd-Here**
# Description
 What does this pull request (PR) do? Why is it necessary? 
 To resolve issues of color attribute changing #5330

## Type of change
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# References
 If this is a bug-fix or otherwise resolves an issue, reference it here with "closes #(issue)" 
 The code is not resolved need approval from @andy-sweet 

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality

### CODE REVIEW

File: napari/_qt/layer_controls/qt_tracks_controls.py [hunk 1]
```diff
@@ -124,10 +124,20 @@ def _on_head_length_change(self):
     def _on_properties_change(self):
         """Change the properties that can be used to color the tracks."""
         with self.layer.events.properties.blocker():
+            pre_store = self.layer.properties_to_color_by
```
Comment (andy-sweet) [inline napari/_qt/layer_controls/qt_tracks_controls.py:127]: `self.layer.properties_to_color_by` returns a list of strings - do we want to store the `self.layer.color_by` string here instead?

File: napari/_qt/layer_controls/qt_tracks_controls.py [hunk 2]
```diff
@@ -124,10 +124,20 @@ def _on_head_length_change(self):
...
+
+            if pre_store in self.layer.properties_to_color_by:
+                self.layers.color_by = pre_store
+                self.color_by_combobox.addItems(
+                    self.layer.properties_to_color_by
+                )
+            else:
+                self.color_by_combobox.addItems(
+                    self.layer.properties_to_color_by
+                )
```
Comment (andy-sweet) [inline napari/_qt/layer_controls/qt_tracks_controls.py:140]: This looks equivalent and shorter/simpler to read.

```suggestion
            if pre_store in self.layer.properties_to_color_by:
                self.layers.color_by = pre_store
            self.color_by_combobox.addItems(self.layer.properties_to_color_by)
```

Looking at it now, I think the issue here is likely being cause by `addItems`, which I think will change the current item of qcombobox to be the first one added here. If that's true then we probably want to block that by using the `qt_signal_blocker` above and then ensuring that `self.layer.color_by` is the current item.

**Review by andy-sweet — COMMENTED** [review]
Thanks for sharing @Kd-Here! I had a couple of comments about the code. I also made a PR on top of this that adds some automated tests to reproduce the underlying the issue here: https://github.com/Kd-Here/napari/pull/1

**Comment by andy-sweet** [comment]
Also, it looks like you committed the changes here straight to the `main` branch on your fork of napari. [As described in the napari guide](https://napari.org/stable/developers/contributing.html#making-changes), the typical/recommended approach is to create a branch on your fork and then create the PR from that to the `main` branch on the main napari repo. That should keep your `main` branch consistent with the main repo's `main`.

Let me know if that doesn't make sense!

**Comment by Kd-Here** [comment]
Hello andy should I close this PR?
And make changes in my personal PR?

**Comment by andy-sweet** [comment]
> Hello andy should I close this PR?

Let's keep this one open for now. We can fix your fork of napari later.

> And make changes in my personal PR?

By personal PR, I think you mean this one right? https://github.com/Kd-Here/napari/pull/1
I created that PR to suggest some automated tests that reproduce the bug described in #5330, but I don't have a fix yet. I can't directly push to your fork, so the PR is a way for me to suggest changes - if the tests look good, feel free to merge them.

I added some more comments on that PR to explain things in a bit more detail.

File: napari/_qt/layer_controls/qt_tracks_controls.py (hunk 2 above)
Comment (andy-sweet) [reply napari/_qt/layer_controls/qt_tracks_controls.py:140]: Another good resource to look at are the Qt docs on `QComboBox`: https://doc.qt.io/qt-5/qcombobox.html#details

Note that the following line in our `qt_tracks_controls.py` means that we react to the `currentTextChanged` signal:

```python
self.color_by_combobox.currentTextChanged.connect(self.change_color_by)
```

And I'm pretty sure that calling `QComboBox.clear` followed by `QComboBox.addItems` (as we do in the code here) means that signal will be emitted (likely making the first item the current one).

**Comment by andy-sweet** [comment]
@Kd-Here : thanks for getting this started! I finished off this work in #5708, so I'm closing this. Your commits are also in that PR, so you should still get credit for your work. Also feel free to add comments to that PR.

**Comment by Kd-Here** [comment]
Thanks @andy-sweet for doing this.
I understand what could I have done better!

**Pull request closed without merging**